File: libs/eave-stdlib-py/src/eave/stdlib/link_handler.py
import asyncio
import enum
import re
from typing import Optional
from urllib.parse import urlparse
from pydantic import UUID4

from eave.stdlib.github_api.operations.content import GetGithubUrlContent
from eave.stdlib.github_api.operations.subscriptions import CreateGithubResourceSubscription

from .core_api.models.subscriptions import SubscriptionInfo

from .eave_origins import EaveApp

class LinkType(enum.StrEnum):
    """
    Link types that we support fetching content from for integration into AI documentation creation.
    """

    github = "github"

# mapping from link type to regex for matching raw links against
SUPPORTED_LINKS: dict[LinkType, list[str]] = {
    LinkType.github: [
        r"github\.com",
        r"github\..+\.com",
    ],
}


def filter_supported_links(urls: list[str]) -> list[tuple[str, LinkType]]:
    supported_links: list[tuple[str, LinkType]] = []
    for link in urls:
        link_type = _get_link_type(link)
        if link_type:
            supported_links.append((link, link_type))
    return supported_links


async def map_url_content(
    origin: EaveApp, eave_team_id: UUID4, urls: list[tuple[str, LinkType]]
) -> list[Optional[str]]:
    """
    Given a list of urls, returns mapping to content found at each link. Order is preserved.

    If an error is encountered while attempting to access the info at a link, the value at
    the position of the link in the returned list is None.
    """
    # gather content from all links in parallel
    tasks: list[asyncio.Task[GetGithubUrlContent.ResponseBody]] = []
    for link, link_type in urls:
        match link_type:
            case LinkType.github:
                tasks.append(
                    asyncio.ensure_future(
                        GetGithubUrlContent.perform(
                            origin=origin,
                            team_id=eave_team_id,
                            input=GetGithubUrlContent.RequestBody(
                                url=link,
                            ),
                        )
                    )
                )

    content_responses: list[Optional[GetGithubUrlContent.ResponseBody]] = await asyncio.gather(*tasks)
    content = list(map(lambda x: x.content if x else None, content_responses))
    return content


async def subscribe_to_file_changes(
    origin: EaveApp, eave_team_id: UUID4, urls: list[tuple[str, LinkType]]
) -> list[SubscriptionInfo]:
    """
    Create Eave Subscriptions to watch for changes in all of the URL resources in `urls`

    eave_team_id -- TeamOrm ID to create the subscription for
    urls -- links paired with their platform type [(url, url platform)]
    returns -- list of subscriptions that got created
    """
    tasks: list[asyncio.Task[Optional[SubscriptionInfo] | Exception]] = []
    for link, link_type in urls:
        tasks.append(
            asyncio.ensure_future(
                _create_subscription_source(origin=origin, url=link, link_type=link_type, eave_team_id=eave_team_id)
            )
        )

    # have asyncio.gather eat any network exceptions and return them as part of result
    completed_tasks: list[Optional[SubscriptionInfo]] = await asyncio.gather(*tasks, return_exceptions=True)
    # only return the successful results
    subscription_sources = [src for src in completed_tasks if type(src) is SubscriptionInfo]
    return subscription_sources


def _get_link_type(link: str) -> Optional[LinkType]:
    """
    Given a link, determine if we support parsing the content from that link.
    Returns link type if supported, otherwise None
    """
    domain = urlparse(link).netloc
    for link_type, regex_patterns in SUPPORTED_LINKS.items():
        if any(re.match(pattern, domain) for pattern in regex_patterns):
            return link_type
    return None


async def _create_subscription_source(
    origin: EaveApp, url: str, link_type: LinkType, eave_team_id: UUID4
) -> Optional[SubscriptionInfo]:
    """
    Insert a subcription into the Eave database to watch the resource at `url`.

    url -- URL resource to create a subscription for watching
    link_type -- resource platform to subscribe on
    eave_team_id -- ID of team to associate subscription with
    """
    # populate required subscription data based on link type
    match link_type:
        case LinkType.github:
            subscription_response = await CreateGithubResourceSubscription.perform(
                origin=origin,
                team_id=eave_team_id,
                input=CreateGithubResourceSubscription.RequestBody(
                    url=url,
                ),
            )
            return SubscriptionInfo(
                subscription=subscription_response.subscription,
                document_reference=subscription_response.document_reference,
            )

    return None

File: libs/eave-stdlib-ts/src/link-handler.ts
import { LinkType } from "./core-api/enums.js";
import { Subscription } from "./core-api/models/subscriptions.js";
import { CreateGithubResourceSubscriptionOperation } from "./github-api/operations/create-subscription.js";
import { GetGithubUrlContentOperation } from "./github-api/operations/get-content.js";
import { eaveLogger } from "./logging.js";
import { RequestArgsTeamId } from "./requests.js";
import { Pair } from "./types.js";

// mapping from link type to regex for matching raw links against
const SUPPORTED_LINKS: { [linkType: string]: Array<RegExp> } = {
  [LinkType.github]: [/github\.com/, /github\..+\.com/],
};

export function filterSupportedLinks(urls: Array<string>): Array<Pair<string, LinkType>> {
  const supportedLinks: Array<Pair<string, LinkType>> = [];
  urls.forEach((link) => {
    const linkType = getLinkType(link);
    if (linkType !== null) {
      supportedLinks.push({ first: link, second: linkType });
    }
  });
  return supportedLinks;
}

/**
 * Given a list of urls, returns mapping to content found at each link. Order is preserved.
 *
 * If an error is encountered while attempting to access the info at a link, the value at
 * the position of the link in the returned list is None.
 */
export async function mapUrlContent({ origin, teamId, urls, ctx }: RequestArgsTeamId & { urls: Array<Pair<string, LinkType>> }): Promise<Array<string | null>> {
  const contentResponses = await Promise.all(
    urls.map(async (linkContext: Pair<string, LinkType>) => {
      const { first: url, second: type } = linkContext;

      switch (type) {
        case LinkType.github: {
          const contentResponse = await GetGithubUrlContentOperation.perform({
            ctx,
            origin,
            teamId,
            input: {
              url,
            },
          });
          return contentResponse.content;
        }
        default:
          eaveLogger.warning(`unsupported link type: ${type}`, ctx);
          return null;
      }
    }),
  );

  return contentResponses;
}

/**
 * Create Eave Subscriptions to watch for changes in all of the URL resources in `urls`
 *
 * @param eaveTeamId -- TeamOrm ID to create the subscription for
 * @param urls -- links paired with their platform type [(url, url platform)]
 */
export async function subscribeToFileChanges({ origin, teamId, urls, ctx }: RequestArgsTeamId & { urls: Array<Pair<string, LinkType>> }): Promise<Array<Subscription>> {
  // string key is a LinkType case, but ts won't let me define it that way
  const subscriptions: Array<Subscription> = [];
  urls.forEach(async (linkContext) => {
    const { first: url, second: linkType } = linkContext;
    const maybeSubscription = await createSubscription({ origin, teamId, url, linkType, ctx });

    if (maybeSubscription !== null) {
      subscriptions.push(maybeSubscription);
    }
  });
  return subscriptions;
}

// smelly conversion from string to enum case to make ts happy
function stringToLinkType(value: string): LinkType | null {
  switch (value) {
    case "github":
      return LinkType.github;
    default:
      return null;
  }
}

/**
 * Given a link, determine if we support parsing the content from that link.
 * @returns link type if supported, otherwise None
 */
function getLinkType(link: string): LinkType | null {
  const domain = new URL(link).hostname;
  let returnValue = null;
  Object.keys(SUPPORTED_LINKS).forEach((linkType) => {
    const patterns = SUPPORTED_LINKS[linkType]!;
    patterns.forEach((pattern) => {
      if (domain.match(pattern)) {
        returnValue = stringToLinkType(linkType);
      }
    });
  });
  return returnValue;
}

/**
 * Insert a subcription to watch the resource at `url` into the Eave database.
 *
 * @param url -- URL resource to create a subscription for watching
 * @param linkType -- resource platform to subscribe on
 * @param eaveTeamId -- ID of team to associate subscription with
 */
async function createSubscription({ origin, teamId, url, linkType, ctx }: RequestArgsTeamId & { url: string; linkType: LinkType }): Promise<Subscription | null> {
  switch (linkType) {
    case LinkType.github: {
      const subscriptionResponse = await CreateGithubResourceSubscriptionOperation.perform({
        ctx,
        origin,
        teamId,
        input: {
          url,
        },
      });
      return subscriptionResponse.subscription;
    }
    default:
      eaveLogger.warning(`unsupported link type: ${linkType}`, ctx);
      return null;
  }
}
